README
NAME
    CPAN::Mini::Visit - A generalised API version of David Golden's
    visitcpan

SYNOPSIS
      CPAN::Mini::Visit->new(
          minicpan => '/minicpan',
          acme     => 0,
          author   => 'ADAMK',
          warnings => 1,
          random   => 1,
          callback => sub {
              print "# counter: $_[0]->{counter}\n";
              print "# archive: $_[0]->{archive}\n";
              print "# tempdir: $_[0]->{tempdir}\n";
              print "# dist:    $_[0]->{dist}\n";
              print "# author:  $_[0]->{author}\n";
          }
      )->run;
      
  # counter: 1234
      # archive: /minicpan/authors/id/A/AD/ADAMK/Config-Tiny-1.00.tar.gz
      # tempdir: /tmp/1a4YRmFAJ3/Config-Tiny-1.00
      # dist:    ADAMK/Config-Tiny-1.00.tar.gz
      # author:  ADAMK

DESCRIPTION
    CPAN::Mini::Extract has been relatively successful at allowing processes
    to run across the contents (or a subset of the contents) of an entire
    minicpan checkout.

    However it has become evident that while it is useful (and theoretically
    optimal from a processing point of view) to maintain an expanded
    minicpan checkout the sheer size of an expanded minicpan is such that it
    becomes an undo burdon to manage, move, copy or even delete a directory
    tree with hundreds of thousands of file totalling in the high single
    gigabytes in size.

    Annoyed by this, David Golden created visitcpan which takes an
    alternative approach of sequentially expanding the tarball of each
    distribution into a temporary directory, do the processing on that
    distribution, and then delete the temporary directory before moving on
    to the next directory.

    This method results in a longer computation time, but with the benefit
    of dramatically reduced system overhead, greater adaptability, and allow
    for easy ad-hoc computations.

    This improvement in flexibility turns out to be worth the extra
    computation time in almost all cases.

    CPAN::Mini::Visit is a simplified and generalised API-based version of
    David Golden's visitcpan script.

    It implements only the process of discovering, iterating and expanding
    archives, before handing off control to an arbitrary callback function
    provided to the constructor.

  new
    Takes a variety of parameters and creates a new visitor object.

    The "minicpan" param should be the root directory of a CPAN::Mini
    download.

    The "callback" param should be a "CODE" reference that will be called
    for each visit. The first parameter passed to the callback will be a
    "HASH" reference containing the tarball location in the "archive" key,
    the location of the temporary directory in the "tempdir" key, the
    canonical CPAN distribution name in the "dist" key, and the author id in
    the "author" key.

    The "acme" param (true by default) can be set to false to exclude any
    distributions that contain the string "Acme", allowing the visit to
    ignore any of the joke modules.

    The "author" param can be provided to limit the visit to only the
    modules owned by a specific author.

    The "random" param will cause the archives to be processed in random
    order if enabled. If not, the archives will be processed in alphabetical
    order.

    The "warnings" param will turn on Archive::Extract warnings if enabled,
    or disable warnings otherwise.

    The "prefer_bin" param will tell Archive::Extract to use binary extract
    instead of CPAN module extract wherever possible. By default, it will
    use module-based extract.

    Returns a CPAN::Mini::Visit object, or throws an exception on error.

  run
    The "run" method executes the visit process, taking no parameters and
    returning true.

    Because the object contains no state information, you may call the "run"
    method multiple times for a single visit object with no ill effects.

SUPPORT
    Bugs should be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CPAN-Mini-Visit>

    For other issues, contact the author.

AUTHOR
    Adam Kennedy <adamk@cpan.org>

COPYRIGHT
    Copyright 2009 - 2011 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Changes for Perl extension CPAN-Mini-Visit

1.14 Fri 11 Nov 2011
	- Moving to a production version.
	- Force dependency on new Locale::Maketext to rid ourself of some
	  troublesome and annoying warnings.

0.13 Wed 30 Mar 2011
	- Clean out mentions of the 'skip' option; this functionality can be 
	  achieved by passing a sub to the 'ignore' option
	- Allow PAUSE ids to include extra characters

0.12_01 Wed 23 Mar 2011
	- Upgrading to Module::Install::DSL 1.00
	- Move sorting into separate sub to allow subclassing (Alexandr Ciornii)
	- Fixed the disfunctional acme => 1 implementation

0.11 Mon 20 Jul 2009
	- Add Harvey to the module blacklist

0.10 Wed  8 Jul 2009
	- Squelch errors resulting from the bad handling of other errors
	- Add support for two levels of warnings

0.09 Wed  8 Jul 2009
	- Try harder to squelch warnings and errors

0.08 Fri 26 Jun 2009
	- Attempt to repair directories with broken permissions

0.07 Fri 26 Jun 2009
	- Fix the warnings flag
	- Add a 'prefer_bin' param to set $Archive::Extract::PREFER_BIN

0.06 Wed 24 Jun 2009
	- Fixing text filters
	- Adding support for the 'random' constructor param

0.05 Tue 23 Jun 2009
	- Including .zip files
	- Including .tar.bz2 files
	- Adding ->warnings options (off by default)
	- Adding explicit skips for three pathalogical files

0.04 Tue  2 Jun 2009
	- Including .tgz files

0.03 Wed 27 May 2009
	- Adding integration with CPAN::Mini

0.02 Sun 17 May 2009
	- Adding customisable filters
	- Adding maximum 3 meg archive size limit

0.01 to be released
	- Creating initial implementation


examples/metadb.pl
#!/usr/bin/perl

use strict;
use File::Remove 'remove';
use File::Spec::Functions ':ALL';
use Parse::CPAN::Meta 'LoadFile';
use CPAN::Mini::Visit;
use DBI;

remove('metadb.sqlite');

my $dbh = DBI->connect('DBI:SQLite:metadb.sqlite');
$dbh->do(<<'END_SQL');
CREATE TABLE meta_dist (
	release TEXT NOT NULL,
	meta_name TEXT,
	meta_version TEXT,
	meta_abstract TEXT,
	meta_generated TEXT,
	meta_from TEXT,
	meta_license TEXT
);
END_SQL

$dbh->do(<<'END_SQL');
CREATE TABLE meta_deps (
	release TEXT NOT NULL,
	phase TEXT NOT NULL,
	module TEXT NOT NULL,
	version TEXT NOT NULL DEFAULT 0
)
END_SQL

$dbh->begin_work;
my $counter   = 0;
my @meta_dist = ();
my @meta_deps = ();
my $visit = CPAN::Mini::Visit->new(
	minicpan => 'D:\minicpan',
	callback => sub {
		print $_[0]->{dist} . "\n";
		my $dist = { release => $_[0]->{dist} };
		my @deps = ();
		my @yaml = eval {
			LoadFile( catfile(
				$_[0]->{tempdir}, 'META.yml',
			) );
		};
		unless ( $@ ) {
			$dist->{meta_name}      = $yaml[0]->{name};
			$dist->{meta_version}   = $yaml[0]->{version};
			$dist->{meta_abstract}  = $yaml[0]->{abstract};
			$dist->{meta_generated} = $yaml[0]->{generated_by};
			$dist->{meta_from}      = $yaml[0]->{version_from};
			$dist->{meta_license}   = $yaml[0]->{license},

			my $requires = $yaml[0]->{requires} || {};
			$requires = { $requires => 0 } unless ref $requires;
			push @deps, map { +{
				release => $_[0]->{dist},
				phase   => 'runtime',
				module  => $_,
				version => $requires->{$_},
			} } sort keys %$requires;

			my $build = $yaml[0]->{build_requires} || {};
			$build = { $build => 0 } unless ref $build;
			push @deps, map { +{
				release => $_[0]->{dist},
				phase   => 'build',
				module  => $_,
				version => $build->{$_},
			} } sort keys %$build;

			my $configure = $yaml[0]->{configure_requires} || {};
			$configure = { $configure => 0 } unless ref $configure;
			push @deps, map { +{
				release => $_[0]->{dist},
				phase   => 'configure',
				module  => $_,
				version => $configure->{$_},
			} } sort keys %$configure;
		}
		$dbh->do(
			'INSERT INTO meta_dist VALUES ( ?, ?, ?, ?, ?, ?, ? )', {},
			$dist->{release},
			$dist->{meta_name},
			$dist->{meta_version},
			$dist->{meta_abstract},
			$dist->{meta_generated},
			$dist->{meta_from},
			$dist->{meta_license},
		);
		foreach ( @deps ) {
			$dbh->do(
				'INSERT INTO meta_deps VALUES ( ?, ?, ?, ? )', {},
				$_->{release},
				$_->{phase},
				$_->{module},
				$_->{version},
			);
		}
		if ( ++$counter % 100 ) {
			$dbh->commit;
			$dbh->begin_work;
		}
	},
)->run;
$dbh->commit;


lib/CPAN/Mini/Visit.pm
package CPAN::Mini::Visit;

=pod

=head1 NAME

CPAN::Mini::Visit - A generalised API version of David Golden's visitcpan

=head1 SYNOPSIS

  CPAN::Mini::Visit->new(
      minicpan => '/minicpan',
      acme     => 0,
      author   => 'ADAMK',
      warnings => 1,
      random   => 1,
      callback => sub {
          print "# counter: $_[0]->{counter}\n";
          print "# archive: $_[0]->{archive}\n";
          print "# tempdir: $_[0]->{tempdir}\n";
          print "# dist:    $_[0]->{dist}\n";
          print "# author:  $_[0]->{author}\n";
      }
  )->run;
  
  # counter: 1234
  # archive: /minicpan/authors/id/A/AD/ADAMK/Config-Tiny-1.00.tar.gz
  # tempdir: /tmp/1a4YRmFAJ3/Config-Tiny-1.00
  # dist:    ADAMK/Config-Tiny-1.00.tar.gz
  # author:  ADAMK

=head1 DESCRIPTION

L<CPAN::Mini::Extract> has been relatively successful at allowing processes
to run across the contents (or a subset of the contents) of an entire
L<minicpan> checkout.

However it has become evident that while it is useful (and theoretically
optimal from a processing point of view) to maintain an expanded minicpan
checkout the sheer size of an expanded minicpan is such that it becomes
an undo burdon to manage, move, copy or even delete a directory tree with
hundreds of thousands of file totalling in the high single gigabytes in size.

Annoyed by this, David Golden created L<visitcpan> which takes an alternative
approach of sequentially expanding the tarball of each distribution into a
temporary directory, do the processing on that distribution, and then delete
the temporary directory before moving on to the next directory.

This method results in a longer computation time, but with the benefit of
dramatically reduced system overhead, greater adaptability, and allow for
easy ad-hoc computations.

This improvement in flexibility turns out to be worth the extra computation
time in almost all cases.

B<CPAN::Mini::Visit> is a simplified and generalised API-based ersion of 
David Golden's L<visitcpan> script.

It implements only the process of discovering, iterating and expanding
archives, before handing off control to an arbitrary callback function
provided to the constructor.

=cut

use 5.008;
use strict;
use warnings;
use Carp                   'croak';
use File::Spec        0.80 ();
use File::Temp        0.21 ();
use File::pushd       1.00 ();
use File::chmod       0.31 ();
use File::Find::Rule  0.27 ();
use Archive::Extract  0.32 ();
use CPAN::Mini       0.576 ();
use Params::Util      1.00 qw{
	_HASH _STRING _ARRAYLIKE _CODELIKE _REGEX
};

our $VERSION = '0.11';

use Object::Tiny 1.06 qw{
	minicpan
	authors
	callback
	acme
	author
	ignore
	random
	warnings
	prefer_bin
};

=pod

=head2 new

Takes a variety of parameters and creates a new visitor object.

The C<minicpan> param should be the root directory of a L<CPAN::Mini>
download.

The C<callback> param should be a C<CODE> reference that will be called
for each visit. The first parameter passed to the callback will be a C<HASH>
reference containing the tarball location in the C<archive> key, the location
of the temporary directory in the C<tempdir> key, the canonical CPAN
distribution name in the C<dist> key, and the author id in the C<author> key.

The C<acme> param (true by default) can be set to false to exclude any
distributions that contain the string "Acme", allowing the visit to ignore
any of the joke modules.

The C<author> param can be provided to limit the visit to only the modules
owned by a specific author.

The C<random> param will cause the archives to be processed in random order
if enabled. If not, the archives will be processed in alphabetical order.

The C<warnings> param will turn on L<Archive::Extract> warnings if enabled,
or disable warnings otherwise.

The C<prefer_bin> param will tell L<Archive::Extract> to use binary extract
instead of CPAN module extract wherever possible. By default, it will use
module-based extract.

Returns a B<CPAN::Mini::Visit> object, or throws an exception on error.

=cut

sub new {
	my $class = shift;
	my $self  = bless { @_ }, $class;

	# Normalise
	$self->{random}     = $self->random     ? 1 : 0;
	$self->{prefer_bin} = $self->prefer_bin ? 1 : 0;
	$self->{warnings}   = 0 unless $self->{warnings};

	# Check params
	unless (
		_HASH($self->minicpan)
		or (
			defined _STRING($self->minicpan)
			and
			-d $self->minicpan
		)
	) {
		croak("Missing or invalid 'minicpan' param");
	}
	unless ( _CODELIKE($self->callback) ) {
		croak("Missing or invalid 'callback' param");
	}
	unless ( defined $self->ignore ) {
		$self->{ignore} = [];
	}
	unless ( _ARRAYLIKE($self->ignore) ) {
		croak("Invalid 'ignore' param");
	}

	# Derive the authors directory
	$self->{authors} = File::Spec->catdir( $self->_minicpan, 'authors', 'id' );
	unless ( -d $self->authors ) {
		croak("Authors directory '$self->{authors}' does not exist");
	}

	return $self;
}

=pod

=head2 run

The C<run> method executes the visit process, taking no parameters and
returning true.

Because the object contains no state information, you may call the C<run>
method multiple times for a single visit object with no ill effects.

=cut

sub run {
	my $self = shift;

	# If we've been passed a HASH minicpan param,
	# do an update_mirror first, before the regular run.
	if ( _HASH($self->minicpan) ) {
		CPAN::Mini->update_mirror(%{$self->minicpan});
	}

	# Search for the files
	my $find  = File::Find::Rule->name('*.tar.gz', '*.tgz', '*.zip', '*.bz2')->file->relative;
	my @files = sort $find->in( $self->authors );
	unless ( $self->acme ) {
		@files = grep { ! /\bAcme\b/ } @files;
	}
	foreach my $filter ( @{$self->ignore} ) {
		if ( defined _STRING($filter) ) {
			$filter = quotemeta $filter;
			$filter = qr/$filter/;
		}
		if ( _REGEX($filter) ) {
			@files = grep { ! /$filter/ } @files;
		} elsif ( _CODELIKE($filter) ) {
			@files = grep { ! $filter->($_) } @files;
		} else {
			die("Missing or invalid filter");
		}
	}

	# Randomise if needed
	if ( $self->random ) {
		@files = sort { rand() <=> rand() } @files;
	}

	# Extract the archive
	my $counter = 0;
	foreach my $file ( @files ) {
		# Derive the main file properties
		my $path = File::Spec->catfile( $self->authors, $file );
		my $dist = $file;
		$dist =~ s|^[A-Z]/[A-Z][A-Z]/|| or die "Bad distpath for $file";
		unless ( $dist =~ /^([A-Z]+)/ ) {
			die "Bad author for $file";
		}
		my $author = "$1";
		if ( $self->author and $self->author ne $author ) {
			next;
		}

		# Explicitly ignore some damaging distributions
		# if we are using Perl extraction
		unless ( $self->prefer_bin ) {
			next if $path =~ /\bHarvey-\d/;
			next if $path =~ /\bText-SenseClusters\b/;
			next if $path =~ /\bBio-Affymetrix\b/;
			next if $path =~ /\bAlien-MeCab\b/;
		}

		# Extract the archive
		local $Archive::Extract::WARN       = !! ($self->warnings > 1);
		local $Archive::Extract::PREFER_BIN = $self->prefer_bin;
		my $archive = Archive::Extract->new( archive => $path );
		my $tmpdir  = File::Temp->newdir;
		my $ok      = 0;
		SCOPE: {
			my $pushd1 = File::pushd::pushd( File::Spec->curdir );
			$ok = eval {
				$archive->extract( to => $tmpdir );
			};
		}
		if ( $@ or not $ok ) {
			if ( $self->warnings > 1 ) {
				warn("Failed to extract '$path': $@");
			} elsif ( $self->warnings ) {
				print "  Failed: $dist\n";
			}
			next;
		}

		# If using bin tools, do an additional check for
		# damaged tarballs with non-executable directories (on unix)
		my $extract = $archive->extract_path;
		unless ( -r $extract and -x $extract ) {
			# Handle special case where we have screwed up
			# permissions on the extract directory.
			# Just assume we have permissions for that.
			File::chmod::chmod( 0755, $extract );
		}

		# Change into the directory
		my $pushd2 = File::pushd::pushd( $extract );

		# Invoke the callback
		$self->callback->( {
			tempdir => $extract,
			archive => $path,
			dist    => $dist,
			author  => $author,
			counter => ++$counter,
		} );
	}

	return 1;
}





######################################################################
# Support Methods

sub _minicpan {
	my $self = shift;
	return _HASH($self->minicpan)
		? $self->minicpan->{local}
		: $self->minicpan;
}

1;

=pod

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CPAN-Mini-Visit>

For other issues, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2009 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
